docs/05-rfcs/2023-01-21-wing-plugins.md

```diff
@@ -0,0 +1,115 @@
+
+Initially the plugin system can be crafted in simple manner, allowing the ability for users to pass in additional `--plugins` option to the wing cli. These plugin files will be Javascript that contains steps for applying non-functional infrastructure requirements.
+
+A wing plugin class consists of the following methods:
+```ts
+  class MyWingPlugin extends WingPlugin {
+    /**
+    * This method hooks in before the wing sdk code is synthesized. 
+    * Allows the plugin access the root app node.
+    */
+    preSynthesis(app) {
+      // do something before synthesis
+    }
+    /**
+    * This method hooks in after the wing sdk code is synthesized.
+    * Allows the plugin access to the resulting terraform config. 
+    */
+    postSynthesis(config) {
+      // do something after synthesis
+    }
+  }
+```
+Additional iterations on the implementation can be made to allow for more complex plugin systems, such as allowing users to define their plugins in a special query language, or implement their plugins using higher level apis.
+
+## Use Cases
+
+The following use cases are examples of real world scenarios that should be solved by implementing a plugin in the plugin system.
+
+- An organization has a service control policy requiring all IAM policy names to be prefixed with a certain string and to include a specific permission boundary. A developer should be able to write wing code and not care about this non-functional requirement. A plugin can be written to apply this requirement to all IAM policies.
+
+- An organization requires its cloud resources to follow strict security requirements to maintain their HIPPA compliance. These rules include encryption at rest, versioning, replication, isolation, and more. These rules can be encapsulated in a plugin that can be applied to all resources.
+
+- An application team is provided a VPC in their account that was provisioned by the platform engineering team. It is required that all the resources in the application are deployed into this VPC.
+
+- Security policies in an organization require the use of customer managed keys for all resources that support encryption at rest. 
+
+- To better manage and understand costs, an organization requires that all resources are tagged with a certain set of tags.
+
+- An organization's security team has decided that all IAM policies should not contain Allow "*" actions. A plugin can be written to scan the synthesized config and throw an error if a policy statement is found, and where it is found.
+
+- A development team is importing an existing stateful resource into their wing code, however the access to the resource should only ever be readonly. The devops team decides to write a plugin to ensure that no resource in the wing code is ever granted a policy that would allow write access to the imported stateful resource.
+
+## Examples
+These examples are meant to be illustrative of the plugin system, and not necessarily the best way to implement the use cases.
```

Comment:
I think the next step of this RFC is to actually try to implement some of the use cases (or all of them). This is how we will validate and demonstrate that the solution can solve the problem.

(if you feel it will be useful, don't hesitate to start implementing or prototype this in the background).